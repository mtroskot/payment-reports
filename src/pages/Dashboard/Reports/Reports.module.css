.container {
  flex: 1;
  margin-right: 5%;
  margin-bottom: 10%;
}

.header {
  flex-direction: row;
  justify-content: space-between;
  height: 100px;
  align-items: center;
}

.content {
  flex: 1;
  justify-content: center;
  align-items: center;
}

.totalBox {
  margin-bottom: 27px;
}

.overviewContainer {
  flex-direction: row;
  height: 100%;
}

.tableContainer {
  flex: 1;
  min-width: 300px;
}
.tableContainerWithPieChart {
  flex: 0.5;
}

.pieChartContainer {
  flex: 0.5;
  margin-left: 20px;
}
