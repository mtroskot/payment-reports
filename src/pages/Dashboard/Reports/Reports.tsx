import React, { useCallback, useMemo, useState } from "react";
import styles from "./Reports.module.css";
import {
  ContentTitle,
  FlexDiv,
  ReportsFilters,
  EmptyContent,
  ReportOverview,
  InfoBox,
  Loader,
  PieChart,
} from "src/components";
import { useTypedTranslation } from "src/hooks/useTypedTranslation";
import {
  DropdownPickerItem,
  Gateway,
  PieChartData,
  Project,
  Report,
  ReportItem,
  ReportOverviewTableHeaderKey,
} from "src/types/interfaces";
import { QueryKey, ReportsFilter } from "src/types/enums";
import { useQuery } from "@tanstack/react-query";
import { getGateways, getProjects, getReport } from "src/services/network/api";
import Colors from "src/styles/colors";
import { formatAmountForDisplay } from "src/utils/number";
import classNames from "classnames";

const Reports = () => {
  const { t } = useTypedTranslation();
  const [activeReportsFilter, setActiveReportsFilter] = useState<ReportsFilter | undefined>();
  const [selectedProject, setSelectedProject] = useState<Project | "all">("all");
  const [selectedGateway, setSelectedGateway] = useState<Gateway | "all">("all");
  const [selectedFromDate, setSelectedFromDate] = useState<Date | undefined>();
  const [selectedToDate, setSelectedToDate] = useState<Date | undefined>();

  const { data: projects = [] } = useQuery<Project[]>([QueryKey.GET_PROJECTS], async () => {
    const response = await getProjects();
    return response.data;
  });

  const { data: gateways = [] } = useQuery<Gateway[]>([QueryKey.GET_GATEWAYS], async () => {
    const response = await getGateways();
    return response.data;
  });

  const {
    data: reports = [],
    isFetching: isLoadingReport = false,
    refetch: refetchReports,
  } = useQuery<Report[]>(
    [QueryKey.GET_REPORTS],
    async () => {
      if (!selectedFromDate || !selectedToDate) {
        return [];
      }

      const response = await getReport({ from: selectedFromDate, to: selectedToDate });
      return response.data;
    },
    { enabled: false }
  );

  const onProjectSelect = useCallback((project: Project | "all") => {
    setActiveReportsFilter(undefined);
    setSelectedProject(project);
  }, []);

  const onGatewaySelect = useCallback((gateway: Gateway | "all") => {
    setActiveReportsFilter(undefined);
    setSelectedGateway(gateway);
  }, []);

  const onFromDateSelect = useCallback(
    (date: Date) => {
      setActiveReportsFilter(undefined);
      setSelectedFromDate(() => {
        if (selectedToDate && date.getTime() > selectedToDate.getTime()) {
          return selectedToDate;
        }
        return date;
      });
    },
    [selectedToDate]
  );

  const onToDateSelect = useCallback(
    (date: Date) => {
      setActiveReportsFilter(undefined);
      setSelectedToDate(() => {
        if (selectedFromDate && selectedFromDate?.getTime() > date.getTime()) {
          return selectedFromDate;
        }
        return date;
      });
    },
    [selectedFromDate]
  );

  const onReportsFilterClick = useCallback((filter: ReportsFilter) => {
    setActiveReportsFilter((prevState) => (prevState === filter ? undefined : filter));
  }, []);

  const onGenerateReportClick = useCallback(() => {
    if (!selectedGateway) {
      setSelectedGateway("all");
    }
    if (!selectedProject) {
      setSelectedProject("all");
    }
    setActiveReportsFilter(undefined);
    refetchReports();
  }, [refetchReports, selectedGateway, selectedProject]);

  const projectPickerItems = useMemo<DropdownPickerItem[]>(
    () => [
      {
        title: t("all_projects"),
        onClick: () => onProjectSelect("all"),
        id: "all",
        isSelected: selectedProject === "all",
      },
      ...projects.map((project) => ({
        title: project.name,
        onClick: () => onProjectSelect(project),
        id: project.projectId,
        isSelected: (selectedProject as Project)?.projectId === project.projectId,
      })),
    ],
    [onProjectSelect, projects, selectedProject, t]
  );

  const gatewayPickerItems = useMemo<DropdownPickerItem[]>(
    () => [
      {
        title: t("all_gateways"),
        onClick: () => onGatewaySelect("all"),
        id: "all",
        isSelected: selectedGateway === "all",
      },
      ...gateways.map((gateway) => ({
        title: gateway.name,
        onClick: () => onGatewaySelect(gateway),
        id: gateway.gatewayId,
        isSelected: (selectedGateway as Gateway)?.gatewayId === gateway.gatewayId,
      })),
    ],
    [gateways, onGatewaySelect, selectedGateway, t]
  );

  const filteredReports = useMemo(() => {
    if (selectedProject === "all" && selectedGateway === "all") {
      return reports;
    }
    return reports.filter((r) => {
      const gatewayCondition = selectedGateway === "all" ? true : selectedGateway.gatewayId === r.gatewayId;
      const projectCondition = selectedProject === "all" ? true : selectedProject.projectId === r.projectId;
      return gatewayCondition && projectCondition;
    });
  }, [reports, selectedGateway, selectedProject]);

  const projectReports = useMemo<ReportItem[]>(() => {
    const projectsWithReports = projects.map((p) => {
      const reportsByProjectId: Report[] = filteredReports.filter((r) => r.projectId === p.projectId);
      return {
        name: p.name,
        id: p.projectId,
        reports: reportsByProjectId,
        total: reportsByProjectId.reduce((acc, currVal) => {
          acc += currVal.amount;
          return acc;
        }, 0),
      };
    });
    if (selectedProject === "all") {
      return projectsWithReports;
    }
    return projectsWithReports.filter((p) => p.id === selectedProject.projectId);
  }, [filteredReports, projects, selectedProject]);

  const gatewayReports = useMemo<ReportItem[]>(() => {
    const gatewaysWithReports = gateways.map((g) => {
      const reportsByGatewayId: Report[] = filteredReports.filter((r) => r.gatewayId === g.gatewayId);
      return {
        name: g.name,
        id: g.gatewayId,
        reports: reportsByGatewayId,
        total: reportsByGatewayId.reduce((acc, currVal) => {
          acc += currVal.amount;
          return acc;
        }, 0),
      };
    });
    if (selectedGateway === "all") {
      return gatewaysWithReports;
    }
    return gatewaysWithReports.filter((g) => g.id === selectedGateway.gatewayId);
  }, [filteredReports, gateways, selectedGateway]);

  const reportData = useMemo<ReadonlyArray<ReportItem>>(() => {
    const showGateWayReports: boolean = selectedProject !== "all" && selectedGateway === "all";
    if (showGateWayReports) {
      return gatewayReports;
    }
    return projectReports;
  }, [gatewayReports, projectReports, selectedGateway, selectedProject]);

  const hasReports = !!selectedGateway && !!selectedProject && reports.length > 0;

  const pieChartData = useMemo<{ total: string; data: PieChartData[] } | null>(() => {
    if (
      (selectedProject === "all" && selectedGateway === "all") ||
      (selectedGateway !== "all" && selectedProject !== "all")
    ) {
      return null;
    }
    return {
      total: `${t(
        selectedProject !== "all" && selectedGateway === "all" ? "gateway_total" : "project_total"
      )} | ${formatAmountForDisplay({
        amount: reportData.reduce((acc, currVal) => {
          return acc + currVal.total;
        }, 0),
      })}`,
      data: reportData.map((gr) => [gr.name, gr.reports.length]),
    };
  }, [reportData, selectedGateway, selectedProject, t]);

  const reportDataTotal = useMemo(
    () =>
      reportData.reduce((acc, currVal) => {
        acc += currVal.total;
        return acc;
      }, 0),
    [reportData]
  );

  const excludedTableData = useMemo<ReportOverviewTableHeaderKey[] | undefined>(() => {
    if (selectedGateway !== "all") {
      return ["gateway"];
    }
    return undefined;
  }, [selectedGateway]);

  return (
    <FlexDiv className={styles.container}>
      <FlexDiv className={styles.header}>
        <ContentTitle title={t("reports")} subtitle={t("easily_generate_report_of_transactions")} />
        <ReportsFilters
          activeDropdown={activeReportsFilter}
          onProjectPickerClick={() => onReportsFilterClick(ReportsFilter.PROJECT)}
          onGatewayPickerClick={() => onReportsFilterClick(ReportsFilter.GATEWAY)}
          onFromDatePickerClick={() => onReportsFilterClick(ReportsFilter.FROM_DATE)}
          onToDatePickerClick={() => onReportsFilterClick(ReportsFilter.TO_DATE)}
          projectPickerItems={projectPickerItems}
          gatewayPickerPickerItems={gatewayPickerItems}
          onFromDateSelect={onFromDateSelect}
          onToDateSelect={onToDateSelect}
          selectedFromDate={selectedFromDate}
          selectedToDate={selectedToDate}
          onGenerateReportClick={onGenerateReportClick}
          selectedProject={selectedProject}
          selectedGateway={selectedGateway}
          generateReportDisabled={false}
          // generateReportDisabled={!selectedFromDate || !selectedToDate}
        />
      </FlexDiv>
      {!hasReports ? (
        isLoadingReport ? (
          <FlexDiv className={styles.content}>
            <Loader color={Colors.SEA_BLUE} />
          </FlexDiv>
        ) : (
          <EmptyContent title={t("no_reports")} description={t("no_reports_description")} />
        )
      ) : (
        <FlexDiv className={styles.overviewContainer}>
          <FlexDiv
            className={classNames(styles.tableContainer, pieChartData ? styles.tableContainerWithPieChart : undefined)}
          >
            <ReportOverview
              displayMode={selectedProject !== "all" && selectedGateway !== "all" ? "table" : "dropdown-table"}
              reports={reportData}
              selectedProjectOption={selectedProject === "all" ? t("all_projects") : selectedProject.name}
              selectedGatewayOption={selectedGateway === "all" ? t("all_gateways") : selectedGateway.name}
              excludedTableData={excludedTableData}
            />
            {!pieChartData && (
              <InfoBox
                className={styles.totalBox}
                title={`${t("total")}: ${formatAmountForDisplay({
                  amount: reportDataTotal,
                })}`}
              />
            )}
          </FlexDiv>
          {pieChartData && (
            <FlexDiv className={styles.pieChartContainer}>
              <PieChart totalText={pieChartData.total} pieChartData={pieChartData.data} />
            </FlexDiv>
          )}
        </FlexDiv>
      )}
    </FlexDiv>
  );
};

export default React.memo(Reports);
