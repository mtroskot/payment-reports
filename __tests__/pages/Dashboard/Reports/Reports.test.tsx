import * as React from "react";
import { render, screen } from "@testing-library/react";
import Reports from "src/pages/Dashboard/Reports";
import { QueryClientTestWrapper } from "__tests__/setup/queryClient";
import userEvent from "@testing-library/user-event";

jest.useFakeTimers().setSystemTime(new Date("2023-01-01"));

describe("renders Reports correctly", () => {
  it("should render empty report state", async () => {
    const { container } = render(
      <QueryClientTestWrapper>
        <Reports />
      </QueryClientTestWrapper>
    );
    expect(screen.getByText("reports")).toBeInTheDocument();
    expect(screen.getByText("easily_generate_report_of_transactions")).toBeInTheDocument();
    expect(screen.getByText("select_project")).toBeInTheDocument();
    expect(screen.getByText("select_gateway")).toBeInTheDocument();
    expect(screen.getByText("from_date")).toBeInTheDocument();
    expect(screen.getByText("to_date")).toBeInTheDocument();
    expect(screen.queryByText("all_gateways")).not.toBeInTheDocument();
    expect(screen.queryByText("all_projects")).not.toBeInTheDocument();
    expect(screen.getByText("generate_report")).toBeInTheDocument();
    expect(screen.getByText("no_reports")).toBeInTheDocument();
    expect(screen.getByText("no_reports_description")).toBeInTheDocument();
    expect(screen.getByTitle("alt_no_reports_icon")).toBeInTheDocument();
    await userEvent.click(screen.getByText("generate_report"));
    expect(container).toMatchSnapshot();
  });
  it("should change filters when selecting them", async () => {
    const { container } = render(
      <QueryClientTestWrapper>
        <Reports />
      </QueryClientTestWrapper>
    );
    await userEvent.click(screen.getByText("from_date"));
    await userEvent.click(screen.getByText("10"));
    await userEvent.click(screen.getByText("to_date"));
    await userEvent.click(screen.getByText("11"));
    expect(await screen.findByText("10/01/2023")).toBeInTheDocument();
    expect(await screen.findByText("11/01/2023")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
