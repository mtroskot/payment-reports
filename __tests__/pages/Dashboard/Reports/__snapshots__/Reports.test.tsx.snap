// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Reports correctly should change filters when selecting them 1`] = `
<div>
  <div
    class=""
  >
    <div
      class=""
    >
      <div
        class=""
      >
        <p
          class=""
        >
          reports
        </p>
        <p
          class=""
          style="line-clamp: 2;"
        >
          easily_generate_report_of_transactions
        </p>
      </div>
      <div
        class=""
      >
        <div
          class=""
        >
          <button
            class=""
            style="background-color: rgb(27, 197, 189); cursor: pointer;"
            type="button"
          >
            <p
              class=""
              style="line-clamp: 1;"
            >
              select_project
            </p>
            <div
              class=""
              title="alt_button_icon"
            />
          </button>
        </div>
        <div
          class=""
        >
          <button
            class=""
            style="background-color: rgb(27, 197, 189); cursor: pointer;"
            type="button"
          >
            <p
              class=""
              style="line-clamp: 1;"
            >
              select_gateway
            </p>
            <div
              class=""
              title="alt_button_icon"
            />
          </button>
        </div>
        <div
          class=""
        >
          <button
            class=""
            style="background-color: rgb(27, 197, 189); cursor: pointer;"
            type="button"
          >
            <p
              class=""
              style="line-clamp: 1;"
            >
              10/01/2023
            </p>
            <div
              class=""
              title="alt_button_icon"
            />
          </button>
        </div>
        <div
          class=""
        >
          <button
            class=""
            style="background-color: rgb(27, 197, 189); cursor: pointer;"
            type="button"
          >
            <p
              class=""
              style="line-clamp: 1;"
            >
              11/01/2023
            </p>
            <div
              class=""
              title="alt_button_icon"
            />
          </button>
        </div>
        <button
          class=""
          style="background-color: rgb(0, 91, 150); cursor: pointer;"
          type="button"
        >
          <p
            class=""
            style="line-clamp: 1;"
          >
            generate_report
          </p>
        </button>
      </div>
    </div>
    <div
      class=""
    >
      <p
        class=""
      >
        no_reports
      </p>
      <p
        class=""
      >
        no_reports_description
      </p>
      <div
        title="alt_no_reports_icon"
      />
    </div>
  </div>
</div>
`;

exports[`renders Reports correctly should render empty report state 1`] = `
<div>
  <div
    class=""
  >
    <div
      class=""
    >
      <div
        class=""
      >
        <p
          class=""
        >
          reports
        </p>
        <p
          class=""
          style="line-clamp: 2;"
        >
          easily_generate_report_of_transactions
        </p>
      </div>
      <div
        class=""
      >
        <div
          class=""
        >
          <button
            class=""
            style="background-color: rgb(27, 197, 189); cursor: pointer;"
            type="button"
          >
            <p
              class=""
              style="line-clamp: 1;"
            >
              select_project
            </p>
            <div
              class=""
              title="alt_button_icon"
            />
          </button>
        </div>
        <div
          class=""
        >
          <button
            class=""
            style="background-color: rgb(27, 197, 189); cursor: pointer;"
            type="button"
          >
            <p
              class=""
              style="line-clamp: 1;"
            >
              select_gateway
            </p>
            <div
              class=""
              title="alt_button_icon"
            />
          </button>
        </div>
        <div
          class=""
        >
          <button
            class=""
            style="background-color: rgb(27, 197, 189); cursor: pointer;"
            type="button"
          >
            <p
              class=""
              style="line-clamp: 1;"
            >
              from_date
            </p>
            <div
              class=""
              title="alt_button_icon"
            />
          </button>
        </div>
        <div
          class=""
        >
          <button
            class=""
            style="background-color: rgb(27, 197, 189); cursor: pointer;"
            type="button"
          >
            <p
              class=""
              style="line-clamp: 1;"
            >
              to_date
            </p>
            <div
              class=""
              title="alt_button_icon"
            />
          </button>
        </div>
        <button
          class=""
          style="background-color: rgb(0, 91, 150); cursor: pointer;"
          type="button"
        >
          <p
            class=""
            style="line-clamp: 1;"
          >
            generate_report
          </p>
        </button>
      </div>
    </div>
    <div
      class=""
    >
      <p
        class=""
      >
        no_reports
      </p>
      <p
        class=""
      >
        no_reports_description
      </p>
      <div
        title="alt_no_reports_icon"
      />
    </div>
  </div>
</div>
`;
